rafthttp/peer.go

```diff
@@ -15,158 +15,149 @@
 package rafthttp
 
 import (
-	"errors"
+	"log"
 	"net/http"
-	"sync"
 	"time"
 
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
 	"github.com/coreos/etcd/etcdserver/stats"
 	"github.com/coreos/etcd/pkg/types"
 	"github.com/coreos/etcd/raft/raftpb"
 )
 
 const (
-	appRespBatchMs = 50
-	propBatchMs    = 10
-
+	DialTimeout      = time.Second
 	ConnReadTimeout  = 5 * time.Second
 	ConnWriteTimeout = 5 * time.Second
+
+	recvBufSize = 4096
 )
 
 type peer struct {
-	sync.Mutex
-
-	id  types.ID
-	cid types.ID
-
-	tr http.RoundTripper
-	// the url this sender post to
-	u  string
-	r  Raft
-	fs *stats.FollowerStats
-
-	batcher     *Batcher
-	propBatcher *ProposalBatcher
-
-	pipeline *pipeline
-	stream   *stream
-
-	paused  bool
-	stopped bool
+	id types.ID
+
+	picker  *streamPicker
+	sendc   chan raftpb.Message
+	recvc   chan raftpb.Message
+	updatec chan string
+	pausec  chan struct{}
+	resumec chan struct{}
+	stopc   chan struct{}
+	done    chan struct{}
 }
 
-func NewPeer(tr http.RoundTripper, u string, id types.ID, cid types.ID, r Raft, fs *stats.FollowerStats, errorc chan error) *peer {
-	return &peer{
-		id:          id,
-		cid:         cid,
-		tr:          tr,
-		u:           u,
-		r:           r,
-		fs:          fs,
-		pipeline:    newPipeline(tr, u, id, cid, fs, errorc),
-		stream:      &stream{},
-		batcher:     NewBatcher(100, appRespBatchMs*time.Millisecond),
-		propBatcher: NewProposalBatcher(100, propBatchMs*time.Millisecond),
+func startPeer(tr http.RoundTripper, u string, local, to, cid types.ID, r Raft, fs *stats.FollowerStats, errorc chan error) *peer {
+	p := &peer{
+		id:      to,
+		picker:  newStreamPicker(fs),
+		sendc:   make(chan raftpb.Message),
+		recvc:   make(chan raftpb.Message, recvBufSize),
+		updatec: make(chan string),
+		pausec:  make(chan struct{}),
+		resumec: make(chan struct{}),
+		stopc:   make(chan struct{}),
+		done:    make(chan struct{}),
 	}
-}
+	go func() {
+		msgAppDialer := newStreamDialer(u, tr, local, to, cid, streamTypeMsgApp)
+		msgAppReader := startStreamReader(msgAppDialer, p.recvc)
+		dialer := newStreamDialer(u, tr, local, to, cid, streamTypeGeneral)
+		reader := startStreamReader(dialer, p.recvc)
+		pipeline := newPipeline(tr, u, to, cid, fs, errorc)
+		var paused bool
+
+		// non-blocking run loop
+		for {
+			select {
+			case m := <-p.sendc:
+				if paused {
+					continue
+				}
+				var writec chan raftpb.Message
+				var name string
+				var size int
+				if writer := p.picker.pick(m); writer != nil {
```

Comment:
i am gonna make the type-specific logic in streamWriter.
picker will become a general stuff, and it will do nothing except pick. gonna combine it with peer.